File: chatbot_logic.py
"""
CryptoBuddy - AI-powered cryptocurrency advisor chatbot logic
"""

import random
from crypto_data import crypto_db

class CryptoBuddy:
    def __init__(self):
        self.name = "CryptoBuddy"
        self.greetings = [
            "Hey there! 👋 Welcome to CryptoBuddy! I'm here to help you navigate the crypto world with smart, sustainable investment advice.",
            "Hello! 🚀 I'm CryptoBuddy, your AI-powered crypto advisor. Let's find you some profitable and eco-friendly investments!",
            "Hi! 🌟 Ready to explore the crypto universe? I'll help you balance profitability with sustainability!"
        ]
        
        self.help_messages = [
            """I can help you with:
• Finding trending cryptocurrencies
• Sustainable investment options  
• Market analysis and advice
• Risk assessment

Try asking: 'Which crypto is most sustainable?' or 'What's trending up?'""",
            """Here's what I can do for you:
• Analyze crypto profitability
• Recommend eco-friendly coins
• Provide market insights
• Balance risk and reward

Just ask me about any crypto or investment strategy!"""
        ]
        
        self.disclaimer = "⚠️  Important: Cryptocurrency investments are highly volatile and risky. Always do your own research and never invest more than you can afford to lose. This advice is for educational purposes only."

    def greet(self):
        """Return a random greeting message"""
        return random.choice(self.greetings)

    def analyze_query(self, user_input):
        """Analyze user query and provide appropriate response"""
        query = user_input.lower().strip()
        
        # Greeting patterns
        if any(word in query for word in ['hello', 'hi', 'hey', 'start', 'beginning']):
            return self.greet()
        
        # Help patterns
        if any(word in query for word in ['help', 'what can you do', 'commands', 'options']):
            return random.choice(self.help_messages)
        
        # Exit patterns
        if any(word in query for word in ['bye', 'goodbye', 'exit', 'quit', 'stop']):
            return "Thanks for using CryptoBuddy! Remember to always do your own research. Happy investing! 🚀"
        
        # Sustainability queries
        if any(word in query for word in ['sustainable', 'eco-friendly', 'green', 'environment', 'energy efficient']):
            return self.get_sustainability_recommendation()
        
        # Profitability queries
        if any(word in query for word in ['profitable', 'rising', 'trending', 'growing', 'bull', 'gains']):
            return self.get_profitability_recommendation()
        
        # Balanced queries
        if any(word in query for word in ['balanced', 'both', 'best overall', 'recommend', 'advice']):
            return self.get_balanced_recommendation()
        
        # Specific crypto queries
        crypto_mentioned = None
        for crypto in crypto_db.keys():
            if crypto.lower() in query or crypto_db[crypto]['symbol'].lower() in query:
                crypto_mentioned = crypto
                break
        
        if crypto_mentioned:
            return self.analyze_specific_crypto(crypto_mentioned)
        
        # Risk and market queries
        if any(word in query for word in ['risk', 'safe', 'volatile', 'market', 'price']):
            return self.get_risk_analysis()
        
        # Default response
        return self.get_default_response()

    def get_sustainability_recommendation(self):
        """Find and recommend the most sustainable cryptocurrency"""
        sustainable_cryptos = [(name, data) for name, data in crypto_db.items() 
                             if data['sustainability_score'] >= 7]
        
        if not sustainable_cryptos:
            return "Currently, I don't have any cryptos with high sustainability scores in my database. Consider looking into proof-of-stake cryptocurrencies which are generally more energy-efficient."
        
        # Sort by sustainability score
        sustainable_cryptos.sort(key=lambda x: x[1]['sustainability_score'], reverse=True)
        top_crypto, data = sustainable_cryptos[0]
        
        response = f"""🌱 **{top_crypto} ({data['symbol']})** is my top sustainable pick!

**Sustainability Score:** {data['sustainability_score']}/10
**Energy Use:** {data['energy_use']}
**Why it's great:** {data['description']}
"""
        
        if data['price_trend'] == 'rising':
            response += "\n📈 Plus, it's currently trending upward!"
        
        response += f"\n\n{self.disclaimer}"
        return response

    def get_profitability_recommendation(self):
        """Find and recommend the most profitable cryptocurrency"""
        profitable_cryptos = [(name, data) for name, data in crypto_db.items() 
                            if data['price_trend'] == 'rising']
        
        if not profitable_cryptos:
            return "The market seems to be in a consolidation phase. Consider dollar-cost averaging into established cryptocurrencies during stable periods."
        
        # Sort by profitability score (considering market cap and trend)
        profitable_cryptos.sort(key=lambda x: self.calculate_profitability_score(x[1]), reverse=True)
        top_crypto, data = profitable_cryptos[0]
        
        response = f"""🚀 **{top_crypto} ({data['symbol']})** is my top profitability pick!

**Trend:** {data['price_trend']} 📈
**Market Cap:** {data['market_cap']}
**Why it's promising:** {data['description']}

{self.disclaimer}"""
        
        return response

    def get_balanced_recommendation(self):
        """Find and recommend the most balanced cryptocurrency"""
        balanced_cryptos = [(name, data, self.calculate_balanced_score(data)) 
                          for name, data in crypto_db.items()]
        
        # Sort by balanced score
        balanced_cryptos.sort(key=lambda x: x[2], reverse=True)
        top_crypto, data, score = balanced_cryptos[0]
        
        response = f"""⚖️ **{top_crypto} ({data['symbol']})** offers the best balance!

**Balanced Score:** {score:.1f}/10
**Trend:** {data['price_trend']} {self.get_trend_emoji(data['price_trend'])}
**Sustainability:** {data['sustainability_score']}/10 {self.get_sustainability_emoji(data['sustainability_score'])}
**Energy Use:** {data['energy_use']}

**Why it's ideal:** {data['description']}

{self.disclaimer}"""
        
        return response

    def analyze_specific_crypto(self, crypto_name):
        """Provide detailed analysis of a specific cryptocurrency"""
        data = crypto_db[crypto_name]
        balanced_score = self.calculate_balanced_score(data)
        
        response = f"""📊 **{crypto_name} ({data['symbol']}) Analysis**

**Price Trend:** {data['price_trend']} {self.get_trend_emoji(data['price_trend'])}
**Market Cap:** {data['market_cap']}
**Sustainability:** {data['sustainability_score']}/10 {self.get_sustainability_emoji(data['sustainability_score'])}
**Energy Use:** {data['energy_use']}
**Overall Score:** {balanced_score:.1f}/10

**About:** {data['description']}

{self.get_recommendation_text(data)}

{self.disclaimer}"""
        
        return response

    def get_risk_analysis(self):
        """Provide general risk analysis for crypto investments"""
        return f"""💡 **Crypto Risk Analysis**

**High Risk/High Reward:** Bitcoin, Ethereum (established but volatile)
**Medium Risk:** Cardano, Solana (growing ecosystems)
**Lower Risk:** Stablecoins, established DeFi tokens

**Key Factors to Consider:**
• Market volatility (can swing 20-50% daily)
• Regulatory changes
• Technology adoption
• Team and community strength

{self.disclaimer}"""

    def get_default_response(self):
        """Default response for unrecognized queries"""
        return """🤔 I'm not sure about that specific query, but I can help you with:

• **"What's sustainable?"** - Eco-friendly crypto recommendations
• **"What's trending?"** - Rising cryptocurrencies  
• **"Best overall?"** - Balanced recommendations
• **"Tell me about [crypto name]"** - Specific analysis

Try asking about Bitcoin, Ethereum, Cardano, or any other crypto!"""

    def calculate_profitability_score(self, data):
        """Calculate profitability score based on trend and market cap"""
        score = 0
        if data['price_trend'] == 'rising':
            score += 4
        elif data['price_trend'] == 'stable':
            score += 2
        
        if data['market_cap'] == 'high':
            score += 3
        elif data['market_cap'] == 'medium':
            score += 2
        else:
            score += 1
        
        return score

    def calculate_balanced_score(self, data):
        """Calculate balanced score considering both profitability and sustainability"""
        profitability = self.calculate_profitability_score(data)
        sustainability = data['sustainability_score']
        return (profitability + sustainability) / 2

    def get_trend_emoji(self, trend):
        """Get emoji for price trend"""
        if trend == 'rising':
            return '📈'
        elif trend == 'stable':
            return '📊'
        else:
            return '📉'

    def get_sustainability_emoji(self, score):
        """Get emoji for sustainability score"""
        if score >= 8:
            return '🌱'
        elif score >= 6:
            return '🌿'
        elif score >= 4:
            return '🌾'
        else:
            return '🔋'

    def get_recommendation_text(self, data):
        """Get recommendation text based on crypto data"""
        balanced_score = self.calculate_balanced_score(data)
        
        if balanced_score >= 7:
            return "✅ **Recommendation:** Strong buy candidate with excellent balance of growth potential and sustainability."
        elif balanced_score >= 5:
            return "⚡ **Recommendation:** Moderate investment opportunity. Consider as part of a diversified portfolio."
        else:
            return "⚠️ **Recommendation:** Higher risk investment. Only consider if you understand the volatility."

File: src/utils/chatbotLogic.ts
import { cryptoDatabase } from '../data/cryptoDatabase';
import { RecommendationResult } from '../types/crypto';

export class CryptoBotLogic {
  private responses = {
    greeting: [
      "Hey there! 👋 Welcome to CryptoBuddy! I'm here to help you navigate the crypto world with smart, sustainable investment advice.",
      "Hello! 🚀 I'm CryptoBuddy, your AI-powered crypto advisor. Let's find you some profitable and eco-friendly investments!",
      "Hi! 🌟 Ready to explore the crypto universe? I'll help you balance profitability with sustainability!"
    ],
    
    help: [
      "I can help you with:\n• Finding trending cryptocurrencies\n• Sustainable investment options\n• Market analysis and advice\n• Risk assessment\n\nTry asking: 'Which crypto is most sustainable?' or 'What's trending up?'",
      "Here's what I can do for you:\n• Analyze crypto profitability\n• Recommend eco-friendly coins\n• Provide market insights\n• Balance risk and reward\n\nJust ask me about any crypto or investment strategy!"
    ],
    
    disclaimer: "⚠️ Important: Cryptocurrency investments are highly volatile and risky. Always do your own research and never invest more than you can afford to lose. This advice is for educational purposes only."
  };

  analyzeQuery(query: string): string {
    const lowerQuery = query.toLowerCase();
    
    // Greeting patterns
    if (this.matchesPattern(lowerQuery, ['hello', 'hi', 'hey', 'start', 'beginning'])) {
      return this.getRandomResponse(this.responses.greeting);
    }
    
    // Help patterns
    if (this.matchesPattern(lowerQuery, ['help', 'what can you do', 'commands', 'options'])) {
      return this.getRandomResponse(this.responses.help);
    }
    
    // Sustainability queries
    if (this.matchesPattern(lowerQuery, ['sustainable', 'eco-friendly', 'green', 'environment', 'energy efficient'])) {
      return this.getSustainabilityRecommendation();
    }
    
    // Profitability queries
    if (this.matchesPattern(lowerQuery, ['profitable', 'rising', 'trending', 'growing', 'bull', 'gains'])) {
      return this.getProfitabilityRecommendation();
    }
    
    // Balanced queries
    if (this.matchesPattern(lowerQuery, ['balanced', 'both', 'best overall', 'recommend', 'advice'])) {
      return this.getBalancedRecommendation();
    }
    
    // Specific crypto queries
    const cryptoMentioned = Object.keys(cryptoDatabase).find(crypto => 
      lowerQuery.includes(crypto.toLowerCase()) || 
      lowerQuery.includes(cryptoDatabase[crypto].symbol.toLowerCase())
    );
    
    if (cryptoMentioned) {
      return this.analyzeCrypto(cryptoMentioned);
    }
    
    // Risk and market queries
    if (this.matchesPattern(lowerQuery, ['risk', 'safe', 'volatile', 'market', 'price'])) {
      return this.getRiskAnalysis();
    }
    
    // Default response
    return this.getDefaultResponse();
  }

  private matchesPattern(query: string, patterns: string[]): boolean {
    return patterns.some(pattern => query.includes(pattern));
  }

  private getRandomResponse(responses: string[]): string {
    return responses[Math.floor(Math.random() * responses.length)];
  }

  private getSustainabilityRecommendation(): string {
    const sustainableCryptos = Object.entries(cryptoDatabase)
      .filter(([_, data]) => data.sustainability_score >= 7)
      .sort(([_, a], [__, b]) => b.sustainability_score - a.sustainability_score);
    
    if (sustainableCryptos.length === 0) {
      return "Currently, I don't have any cryptos with high sustainability scores in my database. Consider looking into proof-of-stake cryptocurrencies which are generally more energy-efficient.";
    }
    
    const [topCrypto, data] = sustainableCryptos[0];
    return `🌱 **${topCrypto} (${data.symbol})** is my top sustainable pick!\n\n` +
           `**Sustainability Score:** ${data.sustainability_score}/10\n` +
           `**Energy Use:** ${data.energy_use}\n` +
           `**Why it's great:** ${data.description}\n\n` +
           `${data.price_trend === 'rising' ? '📈 Plus, it\'s currently trending upward!' : ''}\n\n` +
           this.responses.disclaimer;
  }

  private getProfitabilityRecommendation(): string {
    const profitableCryptos = Object.entries(cryptoDatabase)
      .filter(([_, data]) => data.price_trend === 'rising' && data.market_cap === 'high')
      .sort(([_, a], [__, b]) => this.calculateProfitabilityScore(b) - this.calculateProfitabilityScore(a));
    
    if (profitableCryptos.length === 0) {
      const risingCryptos = Object.entries(cryptoDatabase)
        .filter(([_, data]) => data.price_trend === 'rising');
      
      if (risingCryptos.length === 0) {
        return "The market seems to be in a consolidation phase. Consider dollar-cost averaging into established cryptocurrencies during stable periods.";
      }
      
      const [topCrypto, data] = risingCryptos[0];
      return `📈 **${topCrypto} (${data.symbol})** is showing strong upward momentum!\n\n` +
             `**Trend:** ${data.price_trend}\n` +
             `**Market Cap:** ${data.market_cap}\n` +
             `**Description:** ${data.description}\n\n` +
             this.responses.disclaimer;
    }
    
    const [topCrypto, data] = profitableCryptos[0];
    return `🚀 **${topCrypto} (${data.symbol})** is my top profitability pick!\n\n` +
           `**Trend:** ${data.price_trend}\n` +
           `**Market Cap:** ${data.market_cap}\n` +
           `**Why it's promising:** ${data.description}\n\n` +
           this.responses.disclaimer;
  }

  private getBalancedRecommendation(): string {
    const balancedCryptos = Object.entries(cryptoDatabase)
      .map(([name, data]) => ({
        name,
        data,
        score: this.calculateBalancedScore(data)
      }))
      .sort((a, b) => b.score - a.score);
    
    const top = balancedCryptos[0];
    return `⚖️ **${top.name} (${top.data.symbol})** offers the best balance!\n\n` +
           `**Balanced Score:** ${top.score.toFixed(1)}/10\n` +
           `**Trend:** ${top.data.price_trend}\n` +
           `**Sustainability:** ${top.data.sustainability_score}/10\n` +
           `**Energy Use:** ${top.data.energy_use}\n\n` +
           `**Why it's ideal:** ${top.data.description}\n\n` +
           this.responses.disclaimer;
  }

  private analyzeCrypto(cryptoName: string): string {
    const data = cryptoDatabase[cryptoName];
    const profitabilityScore = this.calculateProfitabilityScore(data);
    const balancedScore = this.calculateBalancedScore(data);
    
    return `📊 **${cryptoName} (${data.symbol}) Analysis**\n\n` +
           `**Price Trend:** ${data.price_trend} ${this.getTrendEmoji(data.price_trend)}\n` +
           `**Market Cap:** ${data.market_cap}\n` +
           `**Sustainability:** ${data.sustainability_score}/10 ${this.getSustainabilityEmoji(data.sustainability_score)}\n` +
           `**Energy Use:** ${data.energy_use}\n` +
           `**Overall Score:** ${balancedScore.toFixed(1)}/10\n\n` +
           `**About:** ${data.description}\n\n` +
           this.getRecommendationText(data) + '\n\n' +
           this.responses.disclaimer;
  }

  private getRiskAnalysis(): string {
    return `💡 **Crypto Risk Analysis**\n\n` +
           `**High Risk/High Reward:** Bitcoin, Ethereum (established but volatile)\n` +
           `**Medium Risk:** Cardano, Solana (growing ecosystems)\n` +
           `**Lower Risk:** Stablecoins, established DeFi tokens\n\n` +
           `**Key Factors to Consider:**\n` +
           `• Market volatility (can swing 20-50% daily)\n` +
           `• Regulatory changes\n` +
           `• Technology adoption\n` +
           `• Team and community strength\n\n` +
           this.responses.disclaimer;
  }

  private getDefaultResponse(): string {
    return `🤔 I'm not sure about that specific query, but I can help you with:\n\n` +
           `• **"What's sustainable?"** - Eco-friendly crypto recommendations\n` +
           `• **"What's trending?"** - Rising cryptocurrencies\n` +
           `• **"Best overall?"** - Balanced recommendations\n` +
           `• **"Tell me about [crypto name]"** - Specific analysis\n\n` +
           `Try asking about Bitcoin, Ethereum, Cardano, or any other crypto!`;
  }

  private calculateProfitabilityScore(data: any): number {
    let score = 0;
    if (data.price_trend === 'rising') score += 4;
    if (data.price_trend === 'stable') score += 2;
    if (data.market_cap === 'high') score += 3;
    if (data.market_cap === 'medium') score += 2;
    if (data.market_cap === 'low') score += 1;
    return score;
  }

  private calculateBalancedScore(data: any): number {
    const profitability = this.calculateProfitabilityScore(data);
    const sustainability = data.sustainability_score;
    return (profitability + sustainability) / 2;
  }

  private getTrendEmoji(trend: string): string {
    switch (trend) {
      case 'rising': return '📈';
      case 'stable': return '📊';
      case 'falling': return '📉';
      default: return '📊';
    }
  }

  private getSustainabilityEmoji(score: number): string {
    if (score >= 8) return '🌱';
    if (score >= 6) return '🌿';
    if (score >= 4) return '🌾';
    return '🔋';
  }

  private getRecommendationText(data: any): string {
    const balancedScore = this.calculateBalancedScore(data);
    
    if (balancedScore >= 7) {
      return `✅ **Recommendation:** Strong buy candidate with excellent balance of growth potential and sustainability.`;
    } else if (balancedScore >= 5) {
      return `⚡ **Recommendation:** Moderate investment opportunity. Consider as part of a diversified portfolio.`;
    } else {
      return `⚠️ **Recommendation:** Higher risk investment. Only consider if you understand the volatility.`;
    }
  }
}

export const cryptoBot = new CryptoBotLogic();